File: src/conventions/pipeline/evaluation/provenance.ts
import type { Call } from "../../../engine/types";
import type { ConditionEvidence } from "../evidence-bundle";

/** Kind of encoder used for call assignment. */
export type EncoderKind = "default-call" | "resolver" | "alternate-encoding" | "frontier-step" | "relay-map";

/** Full decision provenance — why this candidate was selected. */
export interface DecisionProvenance {
  /** Applicability evidence for the winning candidate. */
  readonly applicability: ApplicabilityEvidence;
  /** Module activation traces (which modules were live and why). */
  readonly activation: readonly ActivationTrace[];
  /** Encoding traces (how meanings became concrete calls). */
  readonly encoding: readonly EncodingTrace[];
  /** Legality traces (which calls were legal/illegal). */
  readonly legality: readonly LegalityTrace[];
  /** Arbitration traces (truth-set, ranking inputs per candidate). */
  readonly arbitration: readonly ArbitrationTrace[];
  /** Elimination traces (why candidates were dropped). */
  readonly eliminations: readonly EliminationTrace[];
  /** Handoff traces (module-to-module delegation). */
  readonly handoffs: readonly HandoffTrace[];
}

/** Evidence for why a candidate's conditions were or weren't satisfied. */
export interface ApplicabilityEvidence {
  readonly factDependencies: readonly string[];
  readonly evaluatedConditions: readonly ConditionEvidence[];
}

/** Trace of a candidate eliminated during the pipeline. */
export interface EliminationTrace {
  /** meaningId or callRef of the eliminated candidate. */
  readonly candidateId: string;
  /** Pipeline stage where elimination occurred. */
  readonly stage: "activation" | "applicability" | "encoding" | "legality" | "arbitration";
  /** Human-readable reason for elimination. */
  readonly reason: string;
  /** Supporting evidence (evaluated conditions). */
  readonly evidence: readonly ConditionEvidence[];
  /** How definitive is this elimination? */
  readonly strength: "hard" | "entailed" | "preference";
}

/** Trace of module activation/deactivation. */
export interface ActivationTrace {
  readonly moduleId: string;
  readonly activated: boolean;
  readonly matchedPattern?: string;
  readonly reason?: string;
}

/** Arbitration trace per candidate — truth-set membership, ranking inputs. */
export interface ArbitrationTrace {
  readonly candidateId: string;
  readonly truthSetMember: boolean;
  readonly acceptableSetMember: boolean;
  readonly recommendationRank?: number;
  readonly rankingInputs: {
    readonly recommendationBand: number;
    readonly handFitScore?: number;
    readonly modulePrecedence: number;
    readonly specificity: number;
  };
}

/** How a meaning was encoded into a concrete call. */
export interface EncodingTrace {
  readonly encoderId: string;
  readonly encoderKind: EncoderKind;
  readonly consideredCalls: readonly Call[];
  readonly chosenCall?: Call;
  readonly blockedCalls: readonly { readonly call: Call; readonly reason: string }[];
}

/** Legality check trace for a specific call. */
export interface LegalityTrace {
  readonly call: Call;
  readonly legal: boolean;
  readonly reason?: string;
}

/** Module-to-module handoff trace. */
export interface HandoffTrace {
  readonly fromModuleId: string;
  readonly toModuleId: string;
  readonly reason: string;
}

File: src-tauri/crates/bridge-conventions/src/pipeline/evaluation/provenance.rs
//! Provenance and trace types for pipeline decision tracking.
//!
//! Mirrors TS types from `pipeline/evaluation/provenance.ts`.

use serde::{Deserialize, Serialize};

/// How the encoding was resolved.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum EncoderKind {
    DefaultCall,
    Resolver,
    AlternateEncoding,
    FrontierStep,
    RelayMap,
}

/// Trace for how a surface's encoding was resolved.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct EncodingTrace {
    pub encoder_kind: EncoderKind,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub considered_calls: Option<Vec<String>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub blocked_calls: Option<Vec<String>>,
}

/// Trace for legality check.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct LegalityTrace {
    pub legal: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub reason: Option<String>,
}

/// Trace for why a surface was eliminated.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct EliminationTrace {
    pub gate_id: String,
    pub reason: String,
}

/// Trace of a surface's activation (which module + state entry matched).
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ActivationTrace {
    pub module_id: String,
    pub meaning_id: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub phase: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub state_entry_index: Option<usize>,
}

/// Trace of arbitration decision for a single surface.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ArbitrationTrace {
    pub meaning_id: String,
    pub module_id: String,
    pub outcome: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub reason: Option<String>,
}

/// Trace of negotiation delta handoff.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct HandoffTrace {
    pub meaning_id: String,
    pub module_id: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub delta_fields: Option<Vec<String>>,
}

/// Evidence for how surfaces were found applicable.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct ApplicabilityEvidence {
    pub total_surfaces: usize,
    pub matched_count: usize,
    pub eliminated_count: usize,
}

/// Full decision provenance record.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct DecisionProvenance {
    pub activation: Vec<ActivationTrace>,
    pub arbitration: Vec<ArbitrationTrace>,
    pub handoffs: Vec<HandoffTrace>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub selected_meaning_id: Option<String>,
}
